File: frontend/src/lib/types/threats.ts
export type SecurityThreat = {
  id: string;
  detection_run_id: string | null;
  user_id: string | null;
  threat_type: string;
  category: string;
  severity: string;
  threat_score: number;
  title: string;
  description: string;
  status: string;
  fingerprint: string;
  source_event_type: string | null;
  detected_at: string;
  resolved_at: string | null;
  metadata_json: Record<string, unknown> | null;
};

export type SecurityEventLog = {
  id: string;
  user_id: string | null;
  threat_id: string | null;
  event_type: string;
  threat_type: string | null;
  severity: string;
  message: string;
  payload_json: Record<string, unknown> | null;
  created_at: string;
};

export type ThreatDashboard = {
  open_threats: SecurityThreat[];
  open_total: number;
  by_severity: Record<string, number>;
  by_type: Record<string, number>;
  security_posture: number;
  latest_run: {
    id: string;
    threats_found: number;
    completed_at: string | null;
  } | null;
};

export type UserBehaviorItem = {
  user_id: string;
  threat_count: number;
  avg_threat_score: number;
  risk_level: string;
};

File: backend/app/schemas/threats.py
"""Threat detection API schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class SecurityThreatResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    detection_run_id: UUID | None
    user_id: UUID | None
    threat_type: str
    category: str
    severity: str
    threat_score: int
    title: str
    description: str
    status: str
    fingerprint: str
    source_event_type: str | None
    session_id: UUID | None
    resource_type: str | None
    resource_id: UUID | None
    metadata_json: dict | None
    detected_at: datetime
    resolved_at: datetime | None


class SecurityEventLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID | None
    threat_id: UUID | None
    event_type: str
    threat_type: str | None
    severity: str
    message: str
    payload_json: dict | None
    created_at: datetime


class ThreatDetectionRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    triggered_by_user_id: UUID | None
    threats_found: int
    critical_count: int
    high_count: int
    medium_count: int
    low_count: int
    summary_json: dict | None
    started_at: datetime
    completed_at: datetime | None


class ThreatRunDetailResponse(ThreatDetectionRunResponse):
    threats: list[SecurityThreatResponse] = Field(default_factory=list)


class ThreatDashboardResponse(BaseModel):
    open_threats: list[SecurityThreatResponse]
    open_total: int
    by_severity: dict[str, int]
    by_type: dict[str, int]
    security_posture: int
    latest_run: ThreatDetectionRunResponse | None


class UserBehaviorItem(BaseModel):
    user_id: str
    threat_count: int
    avg_threat_score: float
    risk_level: str


class UserBehaviorResponse(BaseModel):
    items: list[UserBehaviorItem]


class ThreatListResponse(BaseModel):
    items: list[SecurityThreatResponse]
    total: int
    limit: int
    offset: int


class SecurityEventListResponse(BaseModel):
    items: list[SecurityEventLogResponse]
    total: int
    limit: int
    offset: int
